controllers/glanceapi_controller.go

```diff
@@ -289,61 +287,117 @@ func (r *GlanceAPIReconciler) reconcileInit(
+		glanceEndpoints[service.EndpointInternal] = endpoint.Data{
 			Port: glance.GlanceInternalPort,
 		}
 	} else {
-		ports[endpoint.EndpointPublic] = endpoint.Data{
+		glanceEndpoints[service.EndpointPublic] = endpoint.Data{
 			Port: glance.GlancePublicPort,
 		}
 	}
+	apiEndpoints := make(map[string]string)
 
-	for _, metallbcfg := range instance.Spec.ExternalEndpoints {
-		portCfg := ports[metallbcfg.Endpoint]
-		portCfg.MetalLB = &endpoint.MetalLBData{
-			IPAddressPool:   metallbcfg.IPAddressPool,
-			SharedIP:        metallbcfg.SharedIP,
-			SharedIPKey:     metallbcfg.SharedIPKey,
-			LoadBalancerIPs: metallbcfg.LoadBalancerIPs,
+	for endpointType, data := range glanceEndpoints {
+		endpointTypeStr := string(endpointType)
+		endpointName := glance.ServiceName + "-" + endpointTypeStr
+		svcOverride := instance.Spec.Override.Service
+		if svcOverride == nil {
+			svcOverride = &service.RoutedOverrideSpec{}
+		}
+		if svcOverride.EmbeddedLabelsAnnotations == nil {
+			svcOverride.EmbeddedLabelsAnnotations = &service.EmbeddedLabelsAnnotations{}
 		}
 
-		ports[metallbcfg.Endpoint] = portCfg
-	}
+		exportLabels := util.MergeStringMaps(
+			serviceLabels,
+			map[string]string{
+				service.AnnotationEndpointKey: endpointTypeStr,
+			},
+		)
 
-	apiEndpoints, ctrlResult, err := endpoint.ExposeEndpoints(
-		ctx,
-		helper,
-		glance.ServiceName,
-		serviceLabels,
-		ports,
-		time.Duration(5)*time.Second,
-	)
-	if err != nil {
-		instance.Status.Conditions.Set(condition.FalseCondition(
-			condition.ExposeServiceReadyCondition,
-			condition.ErrorReason,
-			condition.SeverityWarning,
-			condition.ExposeServiceReadyErrorMessage,
-			err.Error()))
-		return ctrlResult, err
-	} else if (ctrlResult != ctrl.Result{}) {
-		instance.Status.Conditions.Set(condition.FalseCondition(
-			condition.ExposeServiceReadyCondition,
-			condition.RequestedReason,
-			condition.SeverityInfo,
-			condition.ExposeServiceReadyRunningMessage))
-		return ctrlResult, nil
+		// Create the service
+		svc, err := service.NewService(
+			service.GenericService(&service.GenericServiceDetails{
+				Name:      endpointName,
+				Namespace: instance.Namespace,
+				Labels:    exportLabels,
+				Selector:  serviceLabels,
+				Port: service.GenericServicePort{
+					Name:     endpointName,
+					Port:     data.Port,
+					Protocol: corev1.ProtocolTCP,
+				},
+			}),
+			5,
+			&svcOverride.OverrideSpec,
+		)
+		if err != nil {
+			instance.Status.Conditions.Set(condition.FalseCondition(
+				condition.ExposeServiceReadyCondition,
+				condition.ErrorReason,
+				condition.SeverityWarning,
+				condition.ExposeServiceReadyErrorMessage,
+				err.Error()))
+
+			return ctrl.Result{}, err
+		}
+
+		svc.AddAnnotation(map[string]string{
+			service.AnnotationEndpointKey: endpointTypeStr,
+		})
+
+		// add Annotation to whether creating an ingress is required or not
+		if endpointType == service.EndpointPublic && svc.GetServiceType() == corev1.ServiceTypeClusterIP {
```

Comment:
@stuggi I think I understand the main change here: we create the service at the service operator level, we annotate the service with `AnnotationIngressCreateKey = "core.openstack.org/ingress_create"` in case we need to build an ingress.
E.g. :

```
apiVersion: v1
kind: Service
metadata:
  annotations:
    core.openstack.org/ingress_create: "true"
    core.openstack.org/ingress_name: glance
```

Does the `openstack-operator` watches for the `Service` resources and create the `Routes` for the services that are created with those annotations ?
In the `Glance` use case this would end up having a `Route` only for the `GlanceAPIExternal`, while the internal endpoint is still tied to metalLB ipaddresspool, right?

Reply:
> @stuggi I think I understand the main change here: we create the service at the service operator level, we annotate the service with `AnnotationIngressCreateKey = "core.openstack.org/ingress_create"` in case we need to build an ingress. E.g. :
> 
> ```
> apiVersion: v1
> kind: Service
> metadata:
>   annotations:
>     core.openstack.org/ingress_create: "true"
>     core.openstack.org/ingress_name: glance
> ```
> 
> Does the `openstack-operator` watches for the `Service` resources and create the `Routes` for the services that are created with those annotations ? 

It does not watch the services, it watches the service CR it creates, e.g. glance. When glance updates its status because of condition update, or so, the openstack-operator reconciles and checks the services of glance. For those with the annotation it creates the route and passes in the endpointURL for the service. In a follow up for TLS it then passes the https url to be used for register in keystone. For TLS-E it then will then also create the cert ans passes in the secret name to be used by glance api.

> In the `Glance` use case this would end up having a `Route` only for the `GlanceAPIExternal`, while the internal endpoint is still tied to metalLB ipaddresspool, right?

yes correct. the internal endpoint can be customized using the service override, but for now we don't expect to create a route for it.

Reply:
Ack it works from me, I'm overall +1 with this change, I was trying to properly understand the interaction between the `openstack-operator` (that is supposed to create the `Route`) and the `<Service>`operator, that is supposed to build the service and register endpoints in keystone.